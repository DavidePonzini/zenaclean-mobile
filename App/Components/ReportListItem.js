import React from 'react'
import { Image, StyleSheet, Text, View } from 'react-native'

export default class ReportListItem extends React.Component {
  render () {
    return (
      <View style={styles.container}>
        <Text style={styles.title}>{this.props.item.title}</Text>
        <Text style={styles.descr}>{this.props.item.descr}</Text>
        <Image style={styles.image} source={{ uri: this.props.item.url }} />
      </View>
    )
  }
}

const styles = StyleSheet.create({
  title: {
    fontSize: 30,
    fontWeight: 'bold'
  },
  descr: {

  },
  container: {
    flex: 1,
    borderColor: 'black',
    borderWidth: 1,
    paddingBottom: 5
  },
  image: {
    width: 100,
    height: 100
  }
})
