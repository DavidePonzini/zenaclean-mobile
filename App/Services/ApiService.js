import Secrets from 'react-native-config'

const baseUrl = Secrets.API_URL

const getAddressFromCoords = ({ lat, lng }, cb) => {
  const googleApiUrl = 'https://maps.googleapis.com/maps/api/geocode/json?address='

  fetch(googleApiUrl + lat + ',' + lng + '&key=' + Secrets.GOOGLE_MAPS_API_KEY)
    .then((response) => response.json())
    .then((responseJson) => cb(responseJson))
}

const getMarkers = (cb) => {
  fetch(baseUrl + 'markers.json')
    .then((response) => response.json())
    .then((responseJson) => { cb(responseJson) })
}

const uploadReport = (infoReport) => {
  return fetch(baseUrl + 'markers.json', {
    method: 'POST',
    body: JSON.stringify(infoReport)
  })
}
export default {
  getAddressFromCoords,
  getMarkers,
  uploadReport
}
