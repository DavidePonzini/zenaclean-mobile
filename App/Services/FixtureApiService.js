const SSNlist = ['LNGSVV60T17H509N', 'HTHHLG90R62B112N', 'BRTHZG56P68C982U']
const emailList = ['prova@gmail.com', 'prova@hotmail.it']
export const user1 = {
  email: emailList[0],
  password: '12345678',
  ssn: SSNlist[0],
  id: '5c5afd7c4be7ff1c8d7daacd'
}

const wallets = {}
wallets[user1.id] = 2
export { wallets }
export const fixtureAddress = 'Via del Camoscio'
export const fixtureMarkers = [{ address: fixtureAddress, title: 'Divano abbandonato', latitude: 44.1234, longitude: 8.1234, description: 'Divano abbandonato in via Gramsci', timestamp: '2018-12-09T14:40:35 -02:00', approved_positive: false, approved_negative: false, _id: '5c6580a33a7f647afccd8920', user_id: user1.id, url: 'images/667856780e39f0bbe7407bfdae603f8e', voted_positive: false, voted_negative: false }]
export const fixtureMessageOk = 'Registrazione effettuata con successo!'
export const fixtureMessageNotOk = 'Utente giÃ  esistente'
export const fixtureUserId = 'verylonguserid'
const totVoti = [32, 45, 21]
export let userId = null

const logInUser = (email, password, cb) => {
  return new Promise(resolve => {
    (user1.email === email && user1.password === password) ? resolve({ status: 'ok' }) : resolve({ status: 'no' })
  }).then((res) => {
    if (res.status === 'ok') {
      userId = fixtureUserId
      cb(null, { id: userId, email, eth_address: null })
    } else {
      cb(new Error('login failed'), res)
    }
  })
}

const getAddressFromCoords = ({ lat, lng }, cb) => {
  return new Promise((resolve) => {
    resolve(fixtureAddress)
  }).then((res) => { cb(null, res) })
}

const getMarkers = (neLat, swLat, swLng, neLng, cb) => {
  return new Promise((resolve) => {
    resolve(JSON.stringify(fixtureMarkers))
  }).then((res) => { cb(JSON.parse(res)) })
}

const uploadReport = (infoReport, cb) => {
  return new Promise((resolve) => {
    resolve({ status: 'ok' })
  }).then(() => {
    cb(null)
  })
}

const registerUser = (email, ssn, password, cb) => {
  return new Promise((resolve) => {
    let checked = false
    for (let i = 0; i < SSNlist.length; i++) {
      if (SSNlist[i] === ssn) checked = true
    }
    for (let i = 0; i < emailList.length; i++) {
      if (emailList[i] === email) checked = true
    }
    if (!checked) { resolve(fixtureMessageOk) }
    if (checked) { resolve(fixtureMessageNotOk) }
  }).then((res) => { cb(null, res) })
}

const voteReport = (v, report, cb) => {
  return new Promise((resolve) => {
    let old = totVoti[report]
    let current = totVoti[report] + 1
    if (old < current) resolve({ status: 'ok' })
    else resolve({ status: 'no' })
  }).then((res, current) => {
    if (res.status === 'ok') {
      totVoti[report] = current
      cb(null)
    } else {
      cb(new Error('login failed'), res)
    }
  })
}

const changePassword = (email, oldPassword, newPassword, confirmPassword, cb) => {
  return new Promise((resolve) => {
    if (user1.password === oldPassword && newPassword === confirmPassword) resolve({ status: 'ok' })
    else resolve({ status: 'no' })
  }).then((res) => {
    if (res.status === 'ok') {
      user1.password = newPassword
      cb(null)
    } else {
      cb(new Error('login failed'), res)
    }
  })
}

const isLoggedIn = () => userId != null

const logoutUser = () => {
  userId = null
}
const rehydrateLogin = () => {
  return new Promise((resolve, reject) => {
    if (userId != null) {
      resolve(userId)
    } else {
      reject(new Error('Not logged in'))
    }
  })
}

const getBalance = (eth_addr, cb) => {
  return new Promise(resolve => {
    resolve({ value: wallets[user1.id] })
  }).then(res => {
    cb(res.value)
  })
}

export default {
  getAddressFromCoords,
  getMarkers,
  uploadReport,
  registerUser,
  logInUser,
  changePassword,
  voteReport,
  isLoggedIn,
  logoutUser,
  rehydrateLogin,
  getBalance
}
