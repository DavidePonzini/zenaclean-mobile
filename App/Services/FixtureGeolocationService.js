const fixturePosition = {
  coords: {
    accuracy: 1135,
    altitude: null,
    altitudeAccuracy: null,
    heading: null,
    latitude: 44.4135423,
    longitude: 8.9554223,
    speed: null
  },
  timestamp: 1547641567823
}

const fixtureRegion = {
  latitude: 44.4135423,
  longitude: 8.9554223,
  latitudeDelta: 0.1,
  longitudeDelta: 0.1
}

let lastLocation = fixturePosition
let currentRegion = fixtureRegion

const setCurrentRegion = reg => { currentRegion = reg }
const getCurrentRegion = () => currentRegion

const geolocateOnce = async () => {
  return new Promise(resolve => {
    let pos = fixturePosition
    lastLocation = pos
    resolve(pos)
  })
}

const getLatestPosition = () => lastLocation

const geolocateContinuously = (cb, cbErr, options) => {
  return setInterval(cb, 100)
}

const stopGeolocating = (id) => {
  clearInterval(id)
}

export default {
  geolocateOnce,
  geolocateContinuously,
  stopGeolocating,
  getLatestPosition,
  getCurrentRegion,
  setCurrentRegion
}
