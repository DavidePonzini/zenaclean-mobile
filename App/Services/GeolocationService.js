let lastLocation = null

// initial region
let currentRegion = {
  latitude: 44.4056,
  longitude: 8.9463,
  latitudeDelta: 0.1,
  longitudeDelta: 0.1
}

const setCurrentRegion = reg => { currentRegion = reg }
const getCurrentRegion = () => currentRegion

const geolocateOnce = async () => {
  return new Promise(resolve => {
    navigator.geolocation.getCurrentPosition((pos) => {
      lastLocation = pos
      resolve(pos)
    })
  })
}

const geolocateContinuously = (cb, cbErr, options) => {
  return navigator.geolocation.watchPosition((pos) => {
    cb(pos)
    lastLocation = pos
  }, cbErr, options)
}

const stopGeolocating = (id) => {
  navigator.geolocation.clearWatch(id)
}

const getLatestPosition = () => lastLocation

export default {
  geolocateOnce,
  geolocateContinuously,
  stopGeolocating,
  getLatestPosition,
  getCurrentRegion,
  setCurrentRegion
}
