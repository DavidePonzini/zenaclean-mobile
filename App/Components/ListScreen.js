import React from 'react'
import { FlatList, Text, View } from 'react-native'
import { Button } from 'native-base'
import Icon from 'react-native-vector-icons/FontAwesome'
import { SafeAreaView, NavigationEvents } from 'react-navigation'
import ReportListItem from './ReportListItem'
import api from '../Services/ApiService'
import geolocationService from '../Services/GeolocationService'

export class NoReportsLeftComponent extends React.Component {
  render () {
    return (
      <View>
        <Text>Buone Notizie!</Text>
        <Icon name='street-view' />
        <Text>Non ci sono segnalazioni nella tua zona!</Text>
        <Text>Prenditi un attimo per assaporare il momento oppure </Text>
        <Button><Text>Cerca in un'altra zona</Text></Button>
      </View>)
  }
}

export default class ListScreen extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      markers: [],
      fetching: true
    }
  }
  // this is used by the navigator to set the title in the header
  static navigationOptions = {
    tabBarLabel: 'Lista'
  }

  markerRegionUpdate = () => {
    let region = geolocationService.getCurrentRegion()
    this.region = region
    const neLat = region.latitude + region.latitudeDelta / 2
    const swLat = region.latitude - region.latitudeDelta / 2
    const swLng = region.longitude - region.longitudeDelta / 2
    const neLng = region.longitude + region.longitudeDelta / 2
    this.setState({ fetching: true })
    const that = this
    return api.getMarkers(neLat, swLat, swLng, neLng, (res) => { that.setState({ fetching: false, markers: res }) })
  }
  componentWillUpdate = () => {
    console.log(this.region)
    if (geolocationService.getCurrentRegion() !== this.region) {
      return this.markerRegionUpdate()
    }
  }
  componentWillMount () {
    return this.markerRegionUpdate()
  }

  renderItem = ({ item }) => (
    <ReportListItem
      onPress={(marker) => this.props.navigation.navigate('SingleReport', { marker })}
      marker={item} />)

  render () {
    return (
      <SafeAreaView>
        <NavigationEvents
          onWillFocus={this.markerRegionUpdate} />
        <FlatList
          accessibilityLabel='list-report'
          testID={'list-report'}
          data={this.state.markers}
          onRefresh={this.markerRegionUpdate}
          refreshing={this.state.fetching}
          renderItem={this.renderItem}
          keyExtractor={(_, index) => ('' + index)}
          ListEmptyComponent={!this.state.fetching && <NoReportsLeftComponent />}
        />
      </SafeAreaView>
    )
  }
}
