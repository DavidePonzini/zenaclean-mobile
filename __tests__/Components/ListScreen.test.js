import React from 'react'
import { fixtureMarkers } from '../../App/Services/FixtureApiService'
import { shallow, configure } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import Adapter from 'enzyme-adapter-react-16'
import 'isomorphic-fetch'

import ListScreen from '../../App/Components/ListScreen'
import ReportListItem from '../../App/Components/ReportListItem'
configure({ adapter: new Adapter() })

jest.mock('../../App/Services/ApiService', () => require('../../App/Services/FixtureApiService'))
jest.mock('../../App//Services/GeolocationService', () => require('../../App/Services/FixtureGeolocationService'))
jest.mock('react-navigation', () => { return { SafeAreaView: 'SafeAreaView' } })

const navigation = { navigate: jest.fn() }

describe('List tests', () => {
  const wrapper = shallow(<ListScreen navigation={navigation} />)
  it('displays empty list before markers are loaded', () => {
    const list = wrapper.find('FlatList').first()
    expect(list.props().data).toEqual([])
  })
  it('renders correctly', async () => {
    const instance = wrapper.instance()
    await instance.componentWillMount()
    wrapper.update()
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })
  it('displays NoReportsLeftComponent if no markers are present', () => {
    const instance = wrapper.instance()
    instance.setState({ markers: [] })
    wrapper.update()
    expect(wrapper.find('NoReportsLeftComponent').first()).toBeTruthy()
  })
  it('downloads markers', async () => {
    const instance = wrapper.instance()
    await instance.componentWillMount()
    expect(instance.state.markers).toEqual(fixtureMarkers)
  })
  it('renders markers', () => {
    const element = wrapper.instance().renderItem({ item: fixtureMarkers[0] })
    expect(element.type).toBe(ReportListItem)
    expect(element.props.marker).toEqual(fixtureMarkers[0])
  })
})
