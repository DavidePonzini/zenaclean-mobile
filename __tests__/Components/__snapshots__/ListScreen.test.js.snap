// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List tests renders correctly 1`] = `
<SafeAreaView>
  <Component
    onWillFocus={[Function]}
  />
  <FlatList
    ListEmptyComponent={<NoReportsLeftComponent />}
    accessibilityLabel="list-report"
    data={
      Array [
        Object {
          "address": "Via del Camoscio",
          "description": "Divano abbandonato in via Gramsci",
          "latitude": 44.1234,
          "longitude": 8.1234,
          "timestamp": "2018-12-09T14:40:35 -02:00",
          "title": "Divano abbandonato",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    onRefresh={[Function]}
    refreshing={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    testID="list-report"
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</SafeAreaView>
`;
